File: brokk-code-ts/src/state/promptHistory.ts
import { mkdir, readFile, rename, rm, writeFile } from "node:fs/promises";
import { join } from "node:path";

export const DEFAULT_MAX_HISTORY = 100;

export function historyFile(workspaceDir: string): string {
  return join(workspaceDir, ".brokk", "prompts.json");
}

export async function loadHistory(workspaceDir: string): Promise<string[]> {
  try {
    const raw = await readFile(historyFile(workspaceDir), "utf-8");
    const parsed = JSON.parse(raw);
    if (!Array.isArray(parsed)) {
      return [];
    }
    return parsed.map((item) => String(item));
  } catch {
    return [];
  }
}

export async function saveHistory(workspaceDir: string, history: string[]): Promise<void> {
  const file = historyFile(workspaceDir);
  const stateDir = join(workspaceDir, ".brokk");
  const tmp = `${file}.tmp`;
  await mkdir(stateDir, { recursive: true });
  await writeFile(tmp, `${JSON.stringify(history, null, 2)}\n`, "utf-8");
  await rename(tmp, file);
}

export async function appendPrompt(
  workspaceDir: string,
  prompt: string,
  maxHistory = DEFAULT_MAX_HISTORY
): Promise<void> {
  if (!prompt.trim()) {
    return;
  }
  const history = await loadHistory(workspaceDir);
  history.push(prompt);
  const bounded = history.length > maxHistory ? history.slice(history.length - maxHistory) : history;
  await saveHistory(workspaceDir, bounded);
}

export async function clearHistory(workspaceDir: string): Promise<void> {
  await rm(historyFile(workspaceDir), { force: true });
}

File: brokk-code/brokk_code/prompt_history.py
import json
import logging
from pathlib import Path
from typing import Any, List

logger = logging.getLogger(__name__)

DEFAULT_MAX_HISTORY = 100


def _coerce_path(workspace_dir: Any) -> Path:
    """Robustly coerce workspace_dir or executor objects to a real Path."""
    if workspace_dir is None:
        return Path.cwd()

    # If it's already a Path, return it
    if isinstance(workspace_dir, Path):
        return workspace_dir

    # Handle MagicMocks or objects with workspace_dir attribute
    # We check for __class__.__name__ to avoid triggering mock behavior if possible
    if hasattr(workspace_dir, "workspace_dir"):
        val = workspace_dir.workspace_dir
        if isinstance(val, Path):
            return val
    else:
        val = workspace_dir

    # If it's a mock without workspace_dir, we fall back to CWD to avoid bogus paths
    if "Mock" in type(val).__name__:
        return Path.cwd()

    try:
        if isinstance(val, (str, Path)):
            return Path(val)
    except Exception:
        return Path.cwd()


def get_history_file(workspace_dir: Path) -> Path:
    """Returns the path to the prompt history file in the workspace."""
    return _coerce_path(workspace_dir) / ".brokk" / "prompts.json"


def append_prompt(workspace_dir: Path, prompt: str, max_history: int = DEFAULT_MAX_HISTORY) -> None:
    """Appends a prompt to the history and trims to the last N entries."""
    if not prompt:
        return
    path = _coerce_path(workspace_dir)
    history = load_history(path)
    history.append(prompt)
    if len(history) > max_history:
        history = history[-max_history:]
    save_history(workspace_dir, history)


def load_history(workspace_dir: Path) -> List[str]:
    """Loads the prompt history from the workspace."""
    path = _coerce_path(workspace_dir)
    history_file = get_history_file(path)
    if not history_file.exists():
        return []

    try:
        with history_file.open("r", encoding="utf-8") as f:
            data = json.load(f)
            if isinstance(data, list):
                return [str(item) for item in data]
            return []
    except Exception as e:
        logger.warning(
            "Failed to load prompt history from %s: %s. Starting fresh.", history_file, e
        )
        return []


def save_history(workspace_dir: Path, history: List[str]) -> None:
    """Saves the prompt history to the workspace atomically."""
    path = _coerce_path(workspace_dir)
    history_file = get_history_file(path)
    try:
        history_file.parent.mkdir(parents=True, exist_ok=True)
        temp_file = history_file.with_suffix(".tmp")
        # Ensure history entries are strings to avoid MagicMock serialization issues in tests
        serializable_history = [str(p) for p in history]
        with temp_file.open("w", encoding="utf-8") as f:
            json.dump([str(x) for x in serializable_history], f, indent=4)
        temp_file.replace(history_file)
    except Exception as e:
        logger.error("Failed to save prompt history to %s: %s", history_file, e)


def clear_history(workspace_dir: Path) -> None:
    """Deletes the prompt history file."""
    path = _coerce_path(workspace_dir)
    history_file = get_history_file(path)
    try:
        if history_file.exists():
            history_file.unlink()
    except Exception as e:
        logger.error("Failed to clear prompt history at %s: %s", history_file, e)
